File: src/core/action_parser.py
"""Mock ActionParser: keyword-matching parser for natural-language player input.

Phase 1A: no real LLM calls. Uses keyword matching and simple regex to extract
actions and budgets from player text.
"""

from __future__ import annotations

import re
from typing import Tuple

from src.core.models import ActionPlan, ActionType, PlayerAction, RiskLevel


# ── Keyword → action type mapping ─────────────────────────────────────────────

KEYWORD_MAP: list[Tuple[list[str], ActionType, RiskLevel]] = [
    (["融资", "见投资人", "见投资", "投资人", "路演", "募资", "fundraise", "funding", "vc"], ActionType.FUNDRAISING, RiskLevel.LOW),
    (["招", "hire", "招聘", "雇", "挖人", "团队建设", "扩团队", "招人"], ActionType.TEAM, RiskLevel.MEDIUM),
    (["降价", "投放", "广告", "营销", "推广", "获客", "市场", "增长", "seo", "sem",
      "marketing", "ads", "广告投放", "种子客户", "种子用户"], ActionType.MARKETING, RiskLevel.MEDIUM),
    (["研发", "功能", "产品", "开发", "迭代", "feature", "特性", "技术", "代码",
      "product", "dev", "r&d", "工单", "ai"], ActionType.PRODUCT, RiskLevel.MEDIUM),
    (["转型", "并购", "新市场", "战略", "pivot", "strategy", "收购", "出海",
      "扩张", "新业务"], ActionType.STRATEGY, RiskLevel.HIGH),
]


def _extract_budget(text: str) -> int:
    """Extract a budget number from text. Looks for patterns like '花20万', '50万',
    '预算30', 'budget 100000', '100万元' etc. Returns 0 if none found.
    """
    # Pattern: digits followed by optional 万 (multiply by 10000)
    match = re.findall(r"(\d+)\s*万", text)
    if match:
        return int(match[0]) * 10_000

    # Pattern: bare digits that look like budget amounts (>= 1000)
    bare = re.findall(r"(?:花|预算|花掉|花费|投入|allocate|spend|budget)\s*(\d+)", text)
    if bare:
        return int(bare[0])

    return 0


def _determine_risk(text: str, action_type: ActionType) -> RiskLevel:
    """Determine risk level from text cues."""
    high_keywords = ["激进", "烧钱", "高风险", "all in", "all-in", "豪赌", "猛砸"]
    low_keywords = ["保守", "稳健", "试探", "小规模", "谨慎", "低成本"]

    for kw in high_keywords:
        if kw in text:
            return RiskLevel.HIGH
    for kw in low_keywords:
        if kw in text:
            return RiskLevel.LOW

    # Default risk by action type
    defaults = {
        ActionType.FUNDRAISING: RiskLevel.LOW,
        ActionType.TEAM: RiskLevel.MEDIUM,
        ActionType.MARKETING: RiskLevel.MEDIUM,
        ActionType.PRODUCT: RiskLevel.MEDIUM,
        ActionType.STRATEGY: RiskLevel.HIGH,
    }
    return defaults.get(action_type, RiskLevel.MEDIUM)


def parse(raw_input: str) -> ActionPlan:
    """Parse natural-language player input into an ActionPlan using keyword matching.

    Detects up to 2 distinct action types from the input, extracts budgets,
    and returns a structured ActionPlan.
    """
    actions = []
    seen_types = set()

    for keywords, action_type, default_risk in KEYWORD_MAP:
        if action_type in seen_types:
            continue
        if any(kw in raw_input for kw in keywords):
            budget = _extract_budget(raw_input)
            # If multiple actions would share the same budget, split it
            # For simplicity, first action gets the full extracted budget, others get 0
            if len(actions) > 0 and budget > 0:
                budget = max(0, budget - sum(a.budget for a in actions))

            risk = _determine_risk(raw_input, action_type)
            actions.append(PlayerAction(
                type=action_type,
                intent=raw_input.strip(),
                budget=budget,
                risk_level=risk,
            ))
            seen_types.add(action_type)

        if len(actions) >= 2:
            break

    return ActionPlan(raw_input=raw_input, actions=actions)


def _extract_budget_per_segment(text: str) -> int:
    """Extract budget from a single clause. Looks for 'NN万' pattern and
    returns the amount in 元 (e.g. '30万' → 300000). Returns 0 if none found.
    """
    match = re.search(r'(\d+)万', text)
    if match:
        return int(match.group(1)) * 10_000
    return 0


def parse_multi(raw_input: str) -> ActionPlan:
    """Parse natural-language input into an ActionPlan with up to 5 actions.

    Enhancement over parse():
    - Extracts fundraising (融资+出让) from full text first
    - Splits input by Chinese/English punctuation into clauses
    - Each clause is independently keyword-matched and budget-extracted
    - Skips clauses that contain fundraising-related keywords
    - Max 5 actions total

    Args:
        raw_input: Natural-language player input (e.g. "融资500万出让10%，花200万研发，100万招聘")

    Returns:
        ActionPlan with parsed actions
    """
    actions = []
    seen_types = set()

    # ── Step 1: Extract fundraising from full text ────────────────────────
    fundraise_match = re.search(r'融资(\d+)万.*?出让(\d+)%', raw_input)
    fundraise_amount = 0
    equity_offered = 0
    if fundraise_match:
        fundraise_amount = int(fundraise_match.group(1)) * 10_000
        equity_offered = int(fundraise_match.group(2))

    # ── Step 2: Split into clauses ────────────────────────────────────────
    clauses = re.split(r'[，,；;、]', raw_input)
    clauses = [c.strip() for c in clauses if c.strip()]

    # ── Step 3: Process each clause ───────────────────────────────────────
    for clause in clauses:
        # Skip clauses about fundraising/dilution (already handled)
        if '融资' in clause or '出让' in clause:
            continue

        # Match action type via keywords
        matched_type = None
        matched_risk = RiskLevel.MEDIUM
        for keywords, action_type, default_risk in KEYWORD_MAP:
            if action_type in seen_types:
                continue
            if any(kw in clause for kw in keywords):
                matched_type = action_type
                matched_risk = default_risk
                break

        if matched_type is None:
            continue

        budget = _extract_budget_per_segment(clause)
        risk = _determine_risk(clause, matched_type)
        if risk == RiskLevel.MEDIUM:
            risk = matched_risk

        actions.append(PlayerAction(
            type=matched_type,
            intent=clause.strip(),
            budget=budget,
            risk_level=risk,
        ))
        seen_types.add(matched_type)

        if len(actions) >= 5:
            break

    # ── Step 4: Add fundraising action if detected ────────────────────────
    if fundraise_amount > 0 and ActionType.FUNDRAISING not in seen_types and len(actions) < 5:
        actions.append(PlayerAction(
            type=ActionType.FUNDRAISING,
            intent=f"融资{fundraise_amount // 10_000}万出让{equity_offered}%",
            budget=0,  # fundraising doesn't consume budget
            risk_level=RiskLevel.LOW,
            fundraise_amount=fundraise_amount,
            equity_offered=equity_offered,
        ))

    return ActionPlan(raw_input=raw_input, actions=actions)

File: csharp/StartupSim.Core/Parsing/ActionParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;
using StartupSim.Core.Contracts;

namespace StartupSim.Core.Parsing
{
    public static class ActionParser
    {
        private sealed class KeywordRule
        {
            public KeywordRule(IEnumerable<string> keywords, ActionType actionType, RiskLevel defaultRisk)
            {
                Keywords = keywords.ToArray();
                ActionType = actionType;
                DefaultRisk = defaultRisk;
            }

            public string[] Keywords { get; }
            public ActionType ActionType { get; }
            public RiskLevel DefaultRisk { get; }
        }

        private static readonly KeywordRule[] KeywordMap =
        {
            new KeywordRule(
                new[] { "融资", "见投资人", "见投资", "投资人", "路演", "募资", "fundraise", "funding", "vc" },
                ActionType.Fundraising,
                RiskLevel.Low),
            new KeywordRule(
                new[] { "招", "hire", "招聘", "雇", "挖人", "团队建设", "扩团队", "招人" },
                ActionType.Team,
                RiskLevel.Medium),
            new KeywordRule(
                new[] { "转型", "并购", "新市场", "战略", "策略", "pivot", "strategy", "收购", "出海", "扩张", "新业务" },
                ActionType.Strategy,
                RiskLevel.High),
            new KeywordRule(
                new[] { "降价", "投放", "广告", "营销", "推广", "获客", "市场", "增长", "seo", "sem", "marketing", "ads", "广告投放", "种子客户", "种子用户" },
                ActionType.Marketing,
                RiskLevel.Medium),
            new KeywordRule(
                new[] { "研发", "功能", "产品", "开发", "迭代", "feature", "特性", "技术", "代码", "product", "dev", "r&d", "工单", "ai" },
                ActionType.Product,
                RiskLevel.Medium)
        };

        private static readonly string[] HighRiskKeywords =
        {
            "激进", "烧钱", "高风险", "all in", "all-in", "豪赌", "猛砸"
        };

        private static readonly string[] LowRiskKeywords =
        {
            "保守", "稳健", "试探", "小规模", "谨慎", "低成本"
        };

        public static ActionPlan ParseMulti(string rawInput)
        {
            var input = rawInput ?? string.Empty;
            var actions = new List<PlayerAction>();
            var seenTypes = new HashSet<ActionType>();
            var fundraising = ExtractFundraising(input);

            foreach (var clause in SplitClauses(input))
            {
                if (clause.Contains("融资") || clause.Contains("出让"))
                {
                    continue;
                }

                var matched = MatchRule(clause, seenTypes);
                if (matched == null)
                {
                    continue;
                }

                var explicitRisk = DetermineRisk(clause, matched.ActionType);
                actions.Add(new PlayerAction
                {
                    Type = matched.ActionType,
                    Intent = clause,
                    Budget = ExtractBudgetPerSegment(clause),
                    RiskLevel = explicitRisk == RiskLevel.Medium ? matched.DefaultRisk : explicitRisk
                });
                seenTypes.Add(matched.ActionType);

                if (actions.Count >= 5)
                {
                    break;
                }
            }

            if (fundraising.Amount > 0m && !seenTypes.Contains(ActionType.Fundraising) && actions.Count < 5)
            {
                actions.Add(new PlayerAction
                {
                    Type = ActionType.Fundraising,
                    Intent = $"融资{fundraising.Amount / 10_000m:0}万出让{fundraising.Equity:0}%",
                    Budget = 0m,
                    RiskLevel = RiskLevel.Low,
                    FundraiseAmount = fundraising.Amount,
                    EquityOffered = fundraising.Equity,
                    PostMoneyValuation = fundraising.Equity > 0m
                        ? fundraising.Amount / (fundraising.Equity / 100m)
                        : 0m
                });
            }

            return new ActionPlan
            {
                RawInput = input,
                Actions = actions
            };
        }

        private static IEnumerable<string> SplitClauses(string input)
        {
            return Regex.Split(input, "[，,；;、]")
                .Select(clause => clause.Trim())
                .Where(clause => clause.Length > 0);
        }

        private static KeywordRule? MatchRule(string clause, ISet<ActionType> seenTypes)
        {
            foreach (var rule in KeywordMap)
            {
                if (seenTypes.Contains(rule.ActionType))
                {
                    continue;
                }

                if (rule.Keywords.Any(keyword => ContainsKeyword(clause, keyword)))
                {
                    return rule;
                }
            }

            return null;
        }

        private static bool ContainsKeyword(string text, string keyword)
        {
            return text.IndexOf(keyword, StringComparison.OrdinalIgnoreCase) >= 0;
        }

        private static RiskLevel DetermineRisk(string text, ActionType actionType)
        {
            if (HighRiskKeywords.Any(keyword => ContainsKeyword(text, keyword)))
            {
                return RiskLevel.High;
            }

            if (LowRiskKeywords.Any(keyword => ContainsKeyword(text, keyword)))
            {
                return RiskLevel.Low;
            }

            return actionType == ActionType.Fundraising
                ? RiskLevel.Low
                : actionType == ActionType.Strategy
                    ? RiskLevel.High
                    : RiskLevel.Medium;
        }

        private static decimal ExtractBudgetPerSegment(string text)
        {
            var match = Regex.Match(text, "(\\d+)\\s*万");
            return match.Success ? decimal.Parse(match.Groups[1].Value) * 10_000m : 0m;
        }

        private static (decimal Amount, decimal Equity) ExtractFundraising(string input)
        {
            var amountBeforeEquity = Regex.Match(input, "融资(\\d+)万.*?出让(\\d+)%");
            if (amountBeforeEquity.Success)
            {
                return (
                    decimal.Parse(amountBeforeEquity.Groups[1].Value) * 10_000m,
                    decimal.Parse(amountBeforeEquity.Groups[2].Value));
            }

            var equityBeforeAmount = Regex.Match(input, "出让(\\d+)%.*?融资(\\d+)万");
            if (equityBeforeAmount.Success)
            {
                return (
                    decimal.Parse(equityBeforeAmount.Groups[2].Value) * 10_000m,
                    decimal.Parse(equityBeforeAmount.Groups[1].Value));
            }

            return (0m, 0m);
        }
    }
}
